### PR TITLE
Morpho Gauntlet Prime Strategies for OUSD

### PR DESCRIPTION
**Pull request by naddison36**
# Changes

* `Generalized4626USDTStrategy` added for USDT as Tether's `approve` does not return a bool as per the ERC-20 standard
* Deploy script for Morpho Gauntlet Prime USDC and USDT Strategies to the OUSD Vault
* Fork tests for Morpho Gauntlet Prime USDC and USDT Strategies
* Refactored the exiting MetaMorpho deploy script and fork tests as there will be many Morpho strategies

## Tether USDT support

Tether's `approve` function on USDT is different to a standard ERC-20 token in it does not return a `bool`. It means the following code will fail when calling USDT.

```Solidity
assetToken.approve(platformAddress, type(uint256).max);
```

The usual way to add USDT support is to use Open Zeppelin's `safeApprove` in the `SafeERC20` library. It dynamically handles if the approve call returns data or not. The problem with using `safeApprove` is it checks the allowance is zero before setting a new allowance. That's not a problem for the `Generalized4626Strategy` contract but it does become a problem for the existing `FraxETHStrategy` contract. That's because on initialization it loops through it's two assets (frxETH and WETH) calling `_abstractSetPToken`. The `_abstractSetPToken` function uses the immutable values to approve the asset token to the ERC-4626 Vault which means frxETH is approved twice to the Vault. That fails the `FraxETHStrategy`'s initialization.

The solution used was to implement a new `Generalized4626Strategy` contract that can work for USDT. `Generalized4626USDTStrategy` basically overrides `_approveBase` so the `approve` uses the USDT version with not `bool` return.

```Solidity
interface IUSDT {
    function approve(address _spender, uint _value) external;
}
```

Other solutions considered were:
* Creating a new version of `FraxETHStrategy`
* Calling `safeApprove` twice. The first to set to zero and the next to the max.

## Code Change Checklist

To be completed before internal review begins:

- [ ]  The contract code is complete
- [ ]  Executable deployment file
- [ ]  Fork tests that test after the deployment file runs
- [ ]  Unit tests *if needed
- [ ]  The owner has done a [full checklist review](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) of the code + tests

Internal review:

- [ ] Two approvals by internal reviewers

### CODE REVIEW

**Review by clement-ux — DISMISSED** [review]
The approach for handling USDT `approve` looks good like this to me.

I think there is just a small naming error in this file: [ousd-morpho-guantlet-prime-usdt-usdc.mainnet.fork-test.js](https://github.com/OriginProtocol/origin-dollar/blob/nicka/morpho-stategies/contracts/test/strategies/ousd-morpho-guantlet-prime-usdt-usdc.mainnet.fork-test.js). Shouldn't it be `usdt-usdT.mainnet.fork-test.js` at the end?

Otherwise, everything looks good to me 👍

**Comment by naddison36** [comment]
I fixed the fork test file names

**Review by clement-ux — APPROVED** [review]

**Review by shahthepro — APPROVED** [review]
LGTM

**Comment by DanielVF** [comment]
Newer versions of OZ libs have `forceApprove()` which does handle both USDT and normal ERC20's but does not do the stupid zero check. Unfortunately that's not in the version of OZ libs we have.

File: contracts/utils/addresses.js [hunk 1]
```diff
@@ -171,8 +171,12 @@ addresses.mainnet.MorphoStrategyProxy =
 addresses.mainnet.MorphoAaveStrategyProxy =
   "0x79F2188EF9350A1dC11A062cca0abE90684b0197";
 addresses.mainnet.HarvesterProxy = "0x21Fb5812D70B3396880D30e90D9e5C1202266c89";
-addresses.mainnet.MetaMorphoUSDCSteakHouseVault =
+addresses.mainnet.MorphoSteakhouseUSDCVault =
   "0xBEEF01735c132Ada46AA9aA4c54623cAA92A64CB";
+addresses.mainnet.MorphoGauntletPrimeUSDCVault =
+  "0xdd0f28e19C1780eb6396170735D45153D261490d";
+addresses.mainnet.MorphoGauntletPrimeUSDTVault =
```
Comment (DanielVF) [inline contracts/utils/addresses.js:178]: Verified from morpho UI

File: contracts/utils/addresses.js [hunk 2]
```diff
@@ -171,8 +171,12 @@ addresses.mainnet.MorphoStrategyProxy =
 addresses.mainnet.MorphoAaveStrategyProxy =
   "0x79F2188EF9350A1dC11A062cca0abE90684b0197";
 addresses.mainnet.HarvesterProxy = "0x21Fb5812D70B3396880D30e90D9e5C1202266c89";
-addresses.mainnet.MetaMorphoUSDCSteakHouseVault =
+addresses.mainnet.MorphoSteakhouseUSDCVault =
```
Comment (DanielVF) [inline contracts/utils/addresses.js:174]: Verified from morpho UI

File: contracts/utils/addresses.js [hunk 3]
```diff
@@ -171,8 +171,12 @@ addresses.mainnet.MorphoStrategyProxy =
...
   "0xBEEF01735c132Ada46AA9aA4c54623cAA92A64CB";
+addresses.mainnet.MorphoGauntletPrimeUSDCVault =
+  "0xdd0f28e19C1780eb6396170735D45153D261490d";
```
Comment (DanielVF) [inline contracts/utils/addresses.js:177]: Verified from morpho UI

**Review by DanielVF — APPROVED** [review]

**Pull request merged**